src/main/java/org/zalando/nakadi/validation/schema/SchemaDiff.java

```diff
@@ -0,0 +1,331 @@
+package org.zalando.nakadi.validation.schema;
+
+import org.everit.json.schema.ArraySchema;
+import org.everit.json.schema.CombinedSchema;
+import org.everit.json.schema.EnumSchema;
+import org.everit.json.schema.NumberSchema;
+import org.everit.json.schema.ObjectSchema;
+import org.everit.json.schema.ReferenceSchema;
+import org.everit.json.schema.Schema;
+import org.everit.json.schema.StringSchema;
+import org.zalando.nakadi.domain.SchemaChange;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.Stack;
+import java.util.stream.Collectors;
+
+import static org.everit.json.schema.CombinedSchema.ALL_CRITERION;
+import static org.everit.json.schema.CombinedSchema.ANY_CRITERION;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ADDITIONAL_ITEMS_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ADDITIONAL_PROPERTIES_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ATTRIBUTE_VALUE_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.COMPOSITION_METHOD_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.DEPENDENCY_ARRAY_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.DEPENDENCY_SCHEMA_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.DEPENDENCY_SCHEMA_REMOVED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.DESCRIPTION_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ENUM_ARRAY_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.ID_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.NUMBER_OF_ITEMS_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.PROPERTIES_ADDED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.PROPERTY_REMOVED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.REQUIRED_ARRAY_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.SCHEMA_REMOVED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.SUB_SCHEMA_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.TITLE_CHANGED;
+import static org.zalando.nakadi.domain.SchemaChange.Type.TYPE_CHANGED;
+
+public class SchemaDiff {
```

Comment:
It would break it into smaller classes but the code would be essentially the same.

I thought about other approaches to this problem, but navigating the `Schema` objects is the less error prone in my opinion, since it means that we are inspecting an already interpreted json schema object, and this interpretation is the one from the library we use for validation.

Json schema is full of ambiguities. The everit library puts some opinion on how to interpret that (it even does some kind of "type inference" when some fields are missing). So I would not be willing to move away from it.

Translating these classes to our own domain classes would mean that we would have more than 10 new classes. This fragmentation can lead to even harder to understand code.

This code doesn't look beautiful, but it's surprisingly organized. Every kind of schema has its own method. We could extract these methods to different classes to reduce the overall file size, but I feel it would not improve the overall design.